### PR TITLE
Add C API for printing stack IR

### PR DESCRIPTION
**Pull request by ApsarasX**
I want to write stack IR text form of wasm module to a specific file, but it can only be written to stdout for the lack of function in the existing C API.

So I add the following functions into C API for my needs.

- `BinaryenModulePrintStackIR`: similar to `BinaryenModulePrint`
- `BinaryenModuleWriteStackIR`: similar to `BinaryenModuleWriteText`
- `BinaryenModuleAllocateAndWriteStackIR`: similar to `BinaryenModuleAllocateAndWriteText`

I know my code is not so perfect. If there is a better way, please let me know.

### CODE REVIEW

File: src/binaryen-c.h [hunk 1]
```diff
@@ -2504,6 +2514,12 @@ BinaryenModuleAllocateAndWrite(BinaryenModuleRef module,
 // once not needed anymore.
 BINARYEN_API char* BinaryenModuleAllocateAndWriteText(BinaryenModuleRef module);
 
+// Serialize a module in stack IR form. Implicity allocates the returned
```
Comment (kripken) [inline src/binaryen-c.h:2517]:
```suggestion
// Serialize a module in stack IR form. Implicitly allocates the returned
```

File: src/pass.h [hunk 2]
```diff
@@ -443,6 +444,8 @@ class WalkerPass : public Pass, public WalkerType {
   void setPassRunner(PassRunner* runner_) { runner = runner_; }
 };
 
+std::ostream& printStackIR(std::ostream& o, Module* module);
```
Comment (kripken) [inline src/pass.h:447]: This doesn't seem like the right place for this - it doesn't fit in with anything around it, AFAICT.

I'm not sure offhand what the right place is, but I'd look for similar functions and where they are defined.

Comment (ApsarasX) [reply src/pass.h:447]: Yes, the location of this function is really inappropriate.

But its definition will be used by the C API, and it will also access `GenerateStackIR` and `PrintStackIR`, I can't find a better place for it.

Hope you can find a more suitable location for it.

Comment (kripken) [reply src/pass.h:447]: Hmm, how about at the very end of `wasm-stack.h`? We have similar things at the very end of `wasm.h` for non-stack IR, for comparison.

Comment (ApsarasX) [reply src/pass.h]: Good idea! I have moved printStackIR's declaration into `wasm-stack.h`.

File: src/pass.h [hunk 3]
```diff
@@ -20,6 +20,7 @@
 #include <functional>
 
 #include "mixed_arena.h"
+#include "passes/passes.h"
```
Comment (kripken) [inline src/pass.h:23]: Why is this here?

File: src/passes/Print.cpp [hunk 4]
```diff
@@ -3442,6 +3442,15 @@ printStackIR(StackIR* ir, std::ostream& o, Function* func) {
   return o;
 }
 
+std::ostream& printStackIR(std::ostream& o, Module* module) {
+  wasm::PassRunner runner(module);
+  Pass* generatePass = createGenerateStackIRPass();
```
Comment (kripken) [inline src/passes/Print.cpp:3447]: Instead of this, you can do `runner.add(..)` to create and add the pass, then do `runner.run()`. That would avoid manual memory management here.

File: src/pass.h (hunk 3 above)
Comment (ApsarasX) [reply src/pass.h]: This was originally intended to access `createGenerateStackIRPass`, but is no longer needed.

File: src/passes/Print.cpp (hunk 4 above)
Comment (ApsarasX) [reply src/passes/Print.cpp]: Have modified as you suggested

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
Looks good to land aside from a lint error, see the CI log there. Running `clang-format` should fix it, or you can manually apply the patch from the log.
